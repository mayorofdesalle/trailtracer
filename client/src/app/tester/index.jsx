import '@testing-library/jest-dom';

import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';

import './mocks/i18n.mock';
import './mocks/matchMedia.mock';
import './mocks/resizeObserver.mock';

import store from '@app/store';
import GlobalStyle from '@components/style/GlobalStyle';
import Theme from '@components/style/Theme';

const Providers = ({ children }) => {
    return (
        <Provider store={store}>
            <Theme>
                <GlobalStyle />
                <BrowserRouter>
                    {children}
                </BrowserRouter>
            </Theme>
        </Provider>
    );
};

const customRender = (ui, options) =>
    render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';
export * from 'vitest';

// override render method
export { customRender as render };
