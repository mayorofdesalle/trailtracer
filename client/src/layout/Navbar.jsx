import styled from 'styled-components';
import { Link } from 'react-router-dom';
import logoImg from '../assets/logo.svg';
import Container from '../components/ui/defaults/Container';
import Button from '../components/ui/defaults/Button';
import { RiUser4Line } from 'react-icons/ri';
import { hoverScaling, blurryBackground } from '../components/misc/Mixins';

const NavbarHeader = styled.header`
    width: 100%;
    z-index: 9999;
`;

const NavbarMain = styled(Container)`
    position: fixed;
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
    height: 4rem;
    width: calc(100svw - 2rem - 4px);
    top: calc(1rem + 2px);
    left: calc(1rem + 2px);
    padding: 2rem;

    ${blurryBackground}
`;

const NavbarLogo = styled(Link)`
    display: inline-block;
    height: 1.5rem;
    width: 6.5rem;
    background-color: ${({ theme }) => theme.colors.primary};
    -webkit-mask-image: url(${logoImg});
    mask-image: url(${logoImg});
    mask-size: 6.5rem;
    mask-repeat: no-repeat;
    transition: 200ms ease-in-out;

    ${hoverScaling}

    &:active {
        background-color: ${({ theme }) => theme.colors.secondary};
    }
`;

const LoginButton = styled(Button)`
    text-align: center;
    vertical-align: center;
    color: ${({ theme }) => theme.colors.background};
    background-color: ${({ theme }) => theme.colors.primary};
    font-size: ${({ theme }) => theme.fontSizes.small};
    height: 1.5rem;
    width: 6rem;
    padding: 0.7rem 0.5rem 0.5rem 2rem;
    backdrop-filter: none;
    
    & > * {
        position: absolute;
        transform: translate(-2.5rem, -0.1rem);
        stroke-width: 1px;
        font-size: ${({ theme }) => theme.fontSizes.medium};
    }

    &:active {
        background-color: ${({ theme }) => theme.colors.secondary};
    }
`;

const Navbar = () => {
    return (
        <NavbarHeader id='navigation-bar'>
            <NavbarMain>
                <NavbarLogo to={'/'} />
                <LoginButton><RiUser4Line />SIGN IN</LoginButton>
            </NavbarMain>
        </NavbarHeader>
    );
};

export default Navbar;