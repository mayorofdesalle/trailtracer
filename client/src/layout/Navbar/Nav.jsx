import styled from 'styled-components';

import { glassMorph } from '@components/mixins';

/**
 * Nav
 * @description
 * The styled nav component for the Navbar.
 **/
const Nav = styled.nav`
    display: inline-flex;
    justify-content: space-between;
    align-items: center;
    width: 100%;
    padding: clamp(0.75rem, min(2dvw, 2dvh), 2.5rem);
    z-index: 999;
    ${glassMorph}
    border: none;
`;

export default Nav;