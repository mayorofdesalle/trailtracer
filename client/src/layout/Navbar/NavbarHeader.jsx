import styled from 'styled-components';

import { glassMorph } from '@components/misc/mixins';

/**
 * NavbarHeader
 * @description
 * This is a styled header that is used to contain the navbar.
 **/
const NavbarHeader = styled.header`
    position: absolute;
    display: inline-flex;
    justify-content: space-between;
    align-items: center;
    vertical-align: middle;;
    height: 4svh;
    width: calc(100% - 2rem - 4px);
    top: calc(1rem + 2px);
    left: calc(1rem + 2px);
    padding: 2rem;
    border-radius: 10px;
    z-index: 9999;
    
    ${glassMorph}
`;

export default NavbarHeader;