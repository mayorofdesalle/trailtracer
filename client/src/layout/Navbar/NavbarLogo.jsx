import styled from 'styled-components';

import getImageUrl from '@utils/getImageUrl';
import Button from '@components/ui/Button';

/**
 * NavbarLogo
 * @description
 * This is a styled button that is used as the logo for the navbar.
 **/
const NavbarLogo = styled(Button)`
    border-radius: 0;
    width: clamp(6rem, min(12svw, 12svh), 8rem);
    background-color: ${({ theme }) => theme.colors.primary};
    -webkit-mask-image: url(${getImageUrl('logo')});
    mask-image: url(${getImageUrl('logo')});
    mask-clip: border-box;
    mask-size: contain;
    mask-repeat: no-repeat;
    mask-position: center;
    transition: 200ms ease-in-out;

    &:hover {
        border-radius: 0;
    }

    &:active {
        background-color: ${({ theme }) => theme.colors.secondary};
    }
`;

export default NavbarLogo;