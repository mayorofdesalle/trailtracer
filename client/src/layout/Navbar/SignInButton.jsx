import { useRef } from 'react';
import styled from 'styled-components';

import useHover from '@hooks/useHover';
import Button from '@components/ui/Button';
import Icon from '@components/ui/Icon';

/**
 * ButtonBase
 * @description
 * This component creates the base of the SignInButton.
 **/
const ButtonBase = styled(Button)`
    color: ${({ theme }) => theme.colors.background};
    background-color: ${({ theme }) => theme.colors.primary};
    border-radius: 0.5rem;
    padding: 0 1rem;
    backdrop-filter: none;

    &:active {
        background-color: ${({ theme }) => theme.colors.secondary};
    }

    & > svg {
        transition: 10s;
    }
`;

const SignInButton = () => {
    const signInButton = useRef();
    const isHovered = useHover(signInButton);

    return (
        <ButtonBase ref={signInButton}>
            <Icon name={isHovered ? 'user-fill' : 'user-line'} width={1.5} height={1.5} />
        </ButtonBase>
    );
};

export default SignInButton;