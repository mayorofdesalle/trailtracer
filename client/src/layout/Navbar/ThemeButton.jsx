import { useCallback, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { toggle } from '@features/themeSlice';

import NavbarButton from './NavbarButton';

/**
 * ThemeButton
 * @description
 * Button to toggle between light and dark themes.
 */
const ThemeButton = () => {
    const dispatch = useDispatch();
    const preference = useSelector((state) => state.theme.preference);

    useEffect(() => {
        document.querySelector('meta[name="theme-color"]').setAttribute('content', preference ? '#000000' : '#EBFFF5');
    }, [preference]);

    const toggleTheme = useCallback(() => {
        dispatch(toggle());
    }, [dispatch]);

    return (
        <NavbarButton onClick={toggleTheme} icon={preference ? 'moon' : 'sun'} />
    );
};

export default ThemeButton;