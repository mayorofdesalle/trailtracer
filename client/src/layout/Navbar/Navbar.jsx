import { useCallback } from 'react';
import { useNavigate } from 'react-router-dom';
import styled from 'styled-components';

import Container from '@components/ui/Container';

import Nav from './Nav';
import NavbarButton from './NavbarButton';
import NavbarLogo from './NavbarLogo';
import SettingsButton from './Settings/SettingsButton';
import ExploreBar from './Explore/ExploreBar';

const NavContainer = styled(Container)`
    position: relative;
    width: fit-content;
    height: fit-content;
    margin-left: clamp(0.75rem, min(2dvw, 2dvh), 2.5rem);
    
    & > * {
        margin-left: inherit;

        &:first-child {
            margin-left: 0;
        }
    }
`;

/**
 * Navbar
 * @description
 * Navbar to navigate the site.
 **/
const Navbar = () => {
    const navigate = useNavigate();
    const navigateHome = useCallback(() => navigate('/'), [navigate]);
    const navigateProfile = useCallback(() => navigate('/signin'), [navigate]);

    return (
        <Nav>
            <NavbarLogo onClick={navigateHome} aria-label='Return to homepage' />
            <NavContainer>
                <ExploreBar />
                <NavbarButton onClick={navigateProfile} icon='user' />
                <SettingsButton />
            </NavContainer>
        </Nav>
    );
};

export default Navbar;
