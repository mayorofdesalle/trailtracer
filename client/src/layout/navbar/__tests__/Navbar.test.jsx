import { describe, expect, it, render, screen } from '@app/tester';

import Navbar from '../Navbar';

describe('Navbar', () => {

    it('should render correctly', () => {
        render(<Navbar />);
        const navbar = screen.getByRole('navigation');
        expect(navbar).toBeInTheDocument();
    });

});