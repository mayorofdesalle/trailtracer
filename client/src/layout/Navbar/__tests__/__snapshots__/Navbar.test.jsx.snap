// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navbar > should match snapshot 1`] = `
<DocumentFragment>
  <nav
    class="sc-khjJXk ePvOwI"
  >
    <button
      aria-label="Return to homepage"
      class="sc-beySPh sc-hABBmJ zbkUW lpbBM"
      type="button"
    />
    <div
      class="sc-guDLey sc-dmXWDj kHwwsn jvMJci"
    >
      <form>
        <button
          class="sc-beySPh sc-eTNRI zbkUW chmNhj sc-ktwOfi eXUhpo"
          type="button"
        >
          <svg
            class="sc-jTQCzO ldKJNf"
            fill="currentColor"
            role="img"
          >
            <title>
              menu-search-line
            </title>
            <use
              href="http://localhost:3000/src/assets/icons.svg#menu-search-line"
            />
          </svg>
        </button>
      </form>
      <button
        class="sc-beySPh sc-eTNRI zbkUW chmNhj"
        type="button"
      >
        <svg
          class="sc-jTQCzO ldKJNf"
          fill="currentColor"
          role="img"
        >
          <title>
            user-line
          </title>
          <use
            href="http://localhost:3000/src/assets/icons.svg#user-line"
          />
        </svg>
      </button>
      <button
        class="sc-beySPh sc-eTNRI zbkUW chmNhj"
        name="settingsButton"
        type="button"
      >
        <svg
          class="sc-jTQCzO ldKJNf"
          fill="currentColor"
          role="img"
        >
          <title>
            equalizer-line
          </title>
          <use
            href="http://localhost:3000/src/assets/icons.svg#equalizer-line"
          />
        </svg>
      </button>
    </div>
  </nav>
</DocumentFragment>
`;
